File: generate.js
const fs = require('fs');

function renderDonut(A, B, cols, rows) {
  const chars = '.,-~:;=!*#$@';
  const zbuf = new Float32Array(cols * rows).fill(0);
  const output = new Array(cols * rows).fill(' ');
  const sinA = Math.sin(A), cosA = Math.cos(A);
  const sinB = Math.sin(B), cosB = Math.cos(B);
  const R1 = 1, R2 = 2, K2 = 5;
  const K1 = cols * K2 * 3 / (8 * (R1 + R2));

  for (let theta = 0; theta < Math.PI * 2; theta += 0.07) {
    const sinT = Math.sin(theta), cosT = Math.cos(theta);
    for (let phi = 0; phi < Math.PI * 2; phi += 0.02) {
      const sinP = Math.sin(phi), cosP = Math.cos(phi);
      const cx = R2 + R1 * cosT, cy = R1 * sinT;
      const x = cx * (cosB * cosP + sinA * sinB * sinP) - cy * cosA * sinB;
      const y = cx * (sinB * cosP - sinA * cosB * sinP) + cy * cosA * cosB;
      const z = K2 + cosA * cx * sinP + cy * sinA;
      const ooz = 1 / z;
      const px = Math.floor(cols / 2 + K1 * ooz * x);
      const py = Math.floor(rows / 2 - K1 * ooz * y * 0.5);
      if (px >= 0 && px < cols && py >= 0 && py < rows) {
        const L = cosP * cosT * sinB - cosA * cosT * sinP - sinA * sinT
                + cosB * (cosA * sinT - cosT * sinA * sinP);
        if (L > 0 && ooz > zbuf[py * cols + px]) {
          zbuf[py * cols + px] = ooz;
          output[py * cols + px] = chars[Math.min(Math.floor(L * 8), chars.length - 1)];
        }
      }
    }
  }

  let result = '';
  for (let j = 0; j < rows; j++) {
    result += output.slice(j * cols, j * cols + cols).join('') + '\n';
  }
  return result.trimEnd();
}

function generateSVG() {
  const cols = 44, rows = 22;
  const frameCount = 60;
  const duration = 6; // seconds for full rotation
  const W = 620, H = 340;

  // How tall one frame is in SVG units (px per row * rows)
  const lineHeight = 12;
  const frameH = rows * lineHeight;

  const frames = [];
  for (let f = 0; f < frameCount; f++) {
    const A = (f / frameCount) * Math.PI * 4;
    const B = (f / frameCount) * Math.PI * 2;
    frames.push(renderDonut(A, B, cols, rows));
  }

  // Stack all frames vertically — film strip
  // We duplicate the first frame at the end so the loop is seamless
  const allFrames = [...frames, frames[0]];
  const totalH = allFrames.length * frameH;

  const frameEls = allFrames.map((frame, i) => {
    const yOffset = i * frameH;
    const tspans = frame.split('\n').map((line, j) => {
      const escaped = line
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;');
      return `<tspan x="0" dy="${j === 0 ? '0' : `${lineHeight}px`}">${escaped}</tspan>`;
    }).join('');
    return `<text y="${yOffset + lineHeight}">${tspans}</text>`;
  }).join('\n');

  // Animate translateY from 0 to -frameH*frameCount
  // steps(frameCount, end) snaps between frames with no sub-frame interpolation
  // The duplicated last frame means at t=100% we're back at frame 0 visually
  const totalScroll = frameCount * frameH;

  return `<svg xmlns="http://www.w3.org/2000/svg" width="${W}" height="${H}" viewBox="0 0 ${W} ${H}">
<defs>
  <clipPath id="clip">
    <rect x="0" y="0" width="460" height="${frameH}"/>
  </clipPath>
  <style>
    .strip {
      animation: scroll ${duration}s steps(${frameCount}, end) infinite;
    }
    @keyframes scroll {
      from { transform: translateY(0); }
      to   { transform: translateY(-${totalScroll}px); }
    }
    text {
      font-family: monospace;
      font-size: 10px;
      fill: #e6a030;
      white-space: pre;
    }
    .label {
      font-size: 15px;
      font-weight: bold;
      fill: #e6edf3;
      animation: none;
    }
  </style>
</defs>
<rect width="${W}" height="${H}" fill="#0d1117" rx="12"/>
<text class="label" x="30" y="50">Hi, I'm Thomas</text>
<text class="label" x="340" y="310">I like to make things</text>
<text class="label" x="30" y="310">I like lego too</text>
<g transform="translate(75, 65)" clip-path="url(#clip)">
  <g class="strip">
${frameEls}
  </g>
</g>
</svg>`;
}

fs.writeFileSync('donut.svg', generateSVG());
console.log('donut.svg generated — size:', fs.statSync('donut.svg').size, 'bytes');

File: generate.py
#!/usr/bin/env python3
"""
Generates an animated GIF of a spinning ASCII donut.
Run this locally or in a GitHub Action to produce donut.gif.
"""

import math
import struct
import os
from PIL import Image, ImageDraw, ImageFont

# ── Donut renderer ────────────────────────────────────────────────────────────

CHARS = '.,-~:;=!*#$@'

def render_frame(A, B, cols=44, rows=22):
    R1, R2, K2 = 1.0, 2.0, 5.0
    K1 = cols * K2 * 3 / (8 * (R1 + R2))
    zbuf = [0.0] * (cols * rows)
    out  = [' '] * (cols * rows)
    sinA, cosA = math.sin(A), math.cos(A)
    sinB, cosB = math.sin(B), math.cos(B)

    theta = 0.0
    while theta < math.pi * 2:
        sinT, cosT = math.sin(theta), math.cos(theta)
        phi = 0.0
        while phi < math.pi * 2:
            sinP, cosP = math.sin(phi), math.cos(phi)
            cx = R2 + R1 * cosT
            cy = R1 * sinT
            x = cx * (cosB * cosP + sinA * sinB * sinP) - cy * cosA * sinB
            y = cx * (sinB * cosP - sinA * cosB * sinP) + cy * cosA * cosB
            z = K2 + cosA * cx * sinP + cy * sinA
            ooz = 1.0 / z
            px = int(cols / 2 + K1 * ooz * x)
            py = int(rows / 2 - K1 * ooz * y * 0.5)
            if 0 <= px < cols and 0 <= py < rows:
                L = (cosP * cosT * sinB - cosA * cosT * sinP - sinA * sinT
                     + cosB * (cosA * sinT - cosT * sinA * sinP))
                if L > 0 and ooz > zbuf[py * cols + px]:
                    zbuf[py * cols + px] = ooz
                    idx = min(int(L * 8), len(CHARS) - 1)
                    out[py * cols + px] = CHARS[idx]
            phi += 0.02
        theta += 0.07

    lines = []
    for j in range(rows):
        lines.append(''.join(out[j * cols:(j + 1) * cols]))
    return lines

# ── Image rendering ───────────────────────────────────────────────────────────

BG      = (13, 17, 23)       # #0d1117  GitHub dark background
AMBER   = (255, 105, 180)    # donut colour (pink)
WHITE   = (230, 237, 243)    # label colour

# Canvas size
W, H    = 620, 340

# Donut text block position
DOT_X   = 75
DOT_Y   = 70
CH_W    = 6   # pixels per character (monospace 10 px)
CH_H    = 12  # pixels per line

# Try to load a real monospace font; fall back to default
def load_font(size):
    candidates = [
        '/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf',
        '/usr/share/fonts/truetype/liberation/LiberationMono-Regular.ttf',
        '/System/Library/Fonts/Menlo.ttc',
        'C:/Windows/Fonts/consola.ttf',
    ]
    for path in candidates:
        if os.path.exists(path):
            return ImageFont.truetype(path, size)
    return ImageFont.load_default()

MONO_FONT  = load_font(10)
LABEL_FONT = load_font(14)

LABELS = [
    (30,  32,  "Hi, I'm Thomas"),
    (30,  310, "I like to make things"),
]

def render_image(lines):
    img  = Image.new('RGB', (W, H), BG)
    draw = ImageDraw.Draw(img)

    # Draw rounded rectangle background (PIL >= 8.2)
    try:
        draw.rounded_rectangle([0, 0, W-1, H-1], radius=12, fill=BG)
    except AttributeError:
        draw.rectangle([0, 0, W-1, H-1], fill=BG)

    # Labels
    for x, y, text in LABELS:
        draw.text((x, y), text, font=LABEL_FONT, fill=WHITE)

    # Donut ASCII art
    for row_i, line in enumerate(lines):
        y = DOT_Y + row_i * CH_H
        draw.text((DOT_X, y), line, font=MONO_FONT, fill=AMBER)

    return img

# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    frame_count = 60
    frames = []

    print(f"Rendering {frame_count} frames...")
    for f in range(frame_count):
        A = (f / frame_count) * math.pi * 4
        B = (f / frame_count) * math.pi * 2
        lines = render_frame(A, B)
        img   = render_image(lines)
        frames.append(img)
        if (f + 1) % 10 == 0:
            print(f"  {f + 1}/{frame_count}")

    # duration per frame in centiseconds (100cs = 1s)
    # 60 frames over 6 seconds = 100ms each
    frame_ms = 100  # ms
    frame_cs = frame_ms // 10

    print("Saving donut.gif ...")
    frames[0].save(
        'donut.gif',
        save_all=True,
        append_images=frames[1:],
        duration=frame_ms,
        loop=0,          # 0 = loop forever
        optimize=False,
    )

    size_kb = os.path.getsize('donut.gif') // 1024
    print(f"Done! donut.gif is {size_kb} KB")

if __name__ == '__main__':
    main()